### PR TITLE
fix: branch highlighting was going to end of file for HTML reports

### PR DESCRIPTION
**Pull request by vinteo**
With the logic in place, seems like `startLine` will always == `endLine`
so the condition will never be `false`. Coupled with the fact that
the original `endCol` from the `meta` object seems wrong (always less
than `startCol`), the hightlighting just went to the end of the file.

~~Fix this by removing the redundant logic and always setting the
`endCol` to the end of line using `text.originalLength()`~~

Fix this by removing the redundant logic and setting the
`endCol` to the end of line using `text.originalLength()` when
the given `endCol` is less than `startCol`

### CODE REVIEW

**Comment by xiaofan2406** [comment]
@vinteo According to my test, same fix should be applied to `annotateStatements` as well.

**Comment by bcoe** [comment]
@vinteo mind providing a screen-shot of the highlighting behavior you're seeing?

I wonder if the underlying problem could potentially be a source-map being remapped incorrectly, rather than a problem with the annotator.

before we make significant changes to this somewhat crufty logic, I'd be interested in getting a few tests in place, and I'd also like to make sure that the problem is in fact the reporting logic, and not a badly generated coverage file.

**Comment by bcoe** [comment]
feel free to join the community slack I've setup, if you feel like helping coordinate work around this :) http://devtoolscommunity.herokuapp.com/

**Comment by vinteo** [comment]
@bcoe Screenshot below shows the branch highlighting including lines that have been covered (as you can see by the numbers on the left margin)

image

The underlying problem could be the instrumenter/source maps as the `meta.end.column` for the line seems to be always less than `meta.start.column` (maybe specific to coffeescript), which doesn't make sense.

I used this https://github.com/gotwarlost/istanbul/pull/806 as a starting point, but I feel his fix is not right either as he is mixing columns with lines.

File: packages/istanbul-reports/lib/html/annotator.js [hunk 1]
```diff
@@ -57,13 +57,12 @@ function annotateStatements(fileCoverage, structuredText) {
...
         if (type === 'no' && structuredText[startLine]) {
             if (endLine !== startLine) {
-                endLine = startLine;
                 endCol = structuredText[startLine].text.originalLength();
             }
             text = structuredText[startLine].text;
             text.wrap(startCol,
                 openSpan,
-                startLine === endLine ? endCol : text.originalLength(),
+                startCol < endCol ? endCol : text.originalLength(),
```
Comment (bcoe) [inline packages/istanbul-reports/lib/html/annotator.js]: I think this reads a bit better I like it 👍

Comment (vinteo) [inline packages/istanbul-reports/lib/html/annotator.js]: Yup, this means the existing logic is intact and only fixes the wrong scenario. Although still not ideal it means that if the source maps side of things gets fixed no change is required here.

**Comment by vinteo** [comment]
@bcoe I added some tests, I hope we can get this in 😃

File: packages/istanbul-reports/test/html/annotator.js [hunk 2]
```diff
@@ -32,5 +32,38 @@ describe('annotator', function () {
...
+          return '  var test = "test";';
+        }
+      });
+      annotated.annotatedCode[0].should
+        .equal('<span class="cstat-no" title="statement not covered" >  var test = "test";</span>');
+    });
+
+    // see: https://github.com/istanbuljs/istanbuljs/pull/80
+    it('handles function meta information with end column less than start column', function () {
+      var annotated = annotator.annotateSourceCode(getFixture('github-80b'), {
```
Comment (bcoe) [inline packages/istanbul-reports/test/html/annotator.js]: this is great, I really appreciate the tests.

**Comment by bcoe** [comment]
@vinteo I'm a little concerned about dropping:

`endLine = startLine;`

Mainly because I don't quite know what the original reasoning was behind this logic (it might be attempting to protect against a specific edge-case). Having said this, it seems like you've done a good job of familiarizing yourself with this part of the Istanbul codebase, and I really appreciate the tests.

I say we land this tomorrow, keep an eye out for any bug reports, and prepare to follow up with regression tests. Perhaps we can try to sync up in chat during the day tomorrow to coordinate the rollout?

**Comment by vinteo** [comment]
@bcoe I think its low risk with removing that line as the `endLine` is not used at all further after the changes.  I think the initial intention was to keep the highlighting on the same line which is still maintained.

I'm based in Australia so it might be hard to coordinate a suitable time tomorrow (Saturday too for me) but I am willing to try and work around your schedule. Just ping me on slack.
